Made animation loop interruptible and disposable.

brunnerh opened the pull request:
Dispose now should cancel the animation and calling `animate(null)` should do so as well.

Fixes #13151

brunnerh committed:
Made animation loop interruptible and disposable.

brunnerh commented:
Missed that `isAnimating` has to be reset when breaking the loop via `animate(null)`, otherwise the next call to animate will not work as expected...

brunnerh committed:
Fixed inconsistent isAnimating state.

Mugen87 commented:
I still think it's better to stop the animation loop with `cancelAnimationFrame()` since it's exactly intended for this purpose.

brunnerh commented:
In theory i would agree, but here this would complicate the code. `requestAnimationFrame` is called either on the window or the device depending on state. To properly cancel this we then would have to also keep track of where the frame was requested.

You are of course free to realize this via cancelling; i am not particularly attached to the method by which this is resolved.

mrdoob commented:
I find that logic confusing... is this what you're trying to do?

```js
if ( onAnimationFrame === null ) isAnimating = false;
if ( isAnimating === false ) return;
```

brunnerh commented:
That would be a cleaner way of writing that, yes. The reset of isAnimating was patched in later, after i noticed that it is necessary.

brunnerh committed:
Refactored animation loop exit logic.

brunnerh commented:
Adjusted the code and removed another guard expression that now appears to be unnecessary.

mrdoob commented:
Another approach:

```js
function stop() {

	isAnimating = false;

}

this.animate = function ( callback ) {

	onAnimationFrame = callback;
	onAnimationFrame !== null ? start() : stop();
 
};
```

brunnerh committed:
Refactored requestAnimationFrame logic.
Introduced animation stop function.

brunnerh commented:
That seems sensible. I added that in and also refactored some duplicate code that was in both `start` and `loop`.

Currently in `this.dispose` there is `isAnimating = false` which now could be replaced with `stop()`. I am wondering about the naming, though. `isAnimating = false` seems a lot more descriptive than `stop()`. How about changing the function names to `startAnimation` and `stopAnimation` (or `...Animating`)?

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1050,6 +1050,22 @@ function WebGLRenderer( parameters ) {
+	}
+
+	function stop()
+	{
```
Can you put curly braces in the same line?  😇

mrdoob commented:
> How about changing the function names to `startAnimation` and `stopAnimation` 

Sounds good 👍

Commit:
Animation related function renaming.
Style compliance.

brunnerh commented:
Renamed all related functions, adjusted `dispose` and fixed the braces.

I feel like it would have been more effective for you to just have rejected this and transfer/adjust the code yourself :thinking:.

mrdoob commented:
> I feel like it would have been more effective for you to just have rejected this and transfer/adjust the code yourself :thinking:

Probably. But then you would have lost the credit 🤔

mrdoob commented:
Looks good now!

mrdoob commented:
Thanks!

brunnerh commented:
> Probably. But then you would have lost the credit :thinking:

:man_shrugging:

> Thanks!

:+1:
